config/params/registry.go

```diff
@@ -1,10 +1,162 @@
 package params
 
+import (
+	"github.com/pkg/errors"
+	fieldparams "github.com/prysmaticlabs/prysm/config/fieldparams"
+)
+
+var ErrRegistryCollision = errors.New("registry cannot add config with conflicting fork version schedule")
+var ErrConfigNotFound = errors.New("unable to find requested BeaconChainConfig")
+var ErrCannotNullifyActive = errors.New("cannot set a config marked as active to nil")
+var ErrReplaceNilConfig = errors.New("Replace called with a nil value")
+var ErrConfigNameConflict = errors.New("config with conflicting name already exists")
+
 var Registry *registry
 
 type registry struct {
+	active        *BeaconChainConfig
+	versionToName map[[fieldparams.VersionLength]byte]string
+	nameToConfig  map[string]*BeaconChainConfig
+}
+
+func NewRegistry(configs ...*BeaconChainConfig) *registry {
+	r := &registry{
+		versionToName: make(map[[fieldparams.VersionLength]byte]string),
+		nameToConfig:  make(map[string]*BeaconChainConfig),
+	}
+	for _, c := range configs {
+		if err := r.Add(c); err != nil {
+			panic(err)
+		}
+	}
+	return r
+}
+
+func (r *registry) Add(c *BeaconChainConfig) error {
+	name := c.ConfigName
+	if _, exists := r.nameToConfig[name]; exists {
+		return errors.Wrapf(ErrConfigNameConflict, "ConfigName=%s", name)
+	}
+	c.InitializeForkSchedule()
+	for v := range c.ForkVersionSchedule {
+		if n, exists := r.versionToName[v]; exists {
+			return errors.Wrapf(ErrRegistryCollision, "config name=%s conflicts with existing config named=%s", name, n)
+		}
+		r.versionToName[v] = name
+	}
+	r.nameToConfig[name] = c
+	return nil
+}
+
+func (r *registry) Replace(cfg *BeaconChainConfig) error {
+	if cfg == nil {
+		return ErrReplaceNilConfig
+	}
+	name := cfg.ConfigName
+	r.delete(name)
+	if err := r.Add(cfg); err != nil {
+		return err
+	}
+	if r.active.ConfigName == name {
+		r.active = cfg
+	}
+	return nil
+}
+
+func (r *registry) delete(name string) {
+	c, exists := r.nameToConfig[name]
+	if !exists {
+		return
+	}
+	for v := range c.ForkVersionSchedule {
+		delete(r.versionToName, v)
+	}
+	delete(r.nameToConfig, name)
+}
+
+func (r *registry) ReplaceWithUndo(cfg *BeaconChainConfig) (func() error, error) {
+	name := cfg.ConfigName
+	prev := r.nameToConfig[name]
+	if err := r.Replace(cfg); err != nil {
+		return nil, err
+	}
+	return func() error {
+		if prev == nil {
+			if r.active.ConfigName == name {
+				return errors.Wrapf(ErrCannotNullifyActive, "active config name=%s", name)
+			}
+			r.delete(name)
```

Comment:
Should we be deleting the 'canonical' configs ? 
```go
MainnetConfig(),
		PraterConfig(),
		MinimalSpecConfig(),
		E2ETestConfig(),
		E2EMainnetTestConfig(),
```

If I want temporarily modify our mainnet config with a modified value and then undo it, should we still remove it from the registry ? I understand this function is only meant to be used in tests.

Reply:
Actually nvm, I see that this is added back in `r.Add()`